### PR TITLE
Adding HSLColor and color 'within' matchers for HSVColor and HSLColor

### PR DESCRIPTION
**Pull request by gspencergoog**
This adds an HSLColor class which uses a perceptual color space based upon human perception of colored light (as opposed to HSV, which is based on pigment colors).

You can see the difference in the color spaces [here](https://en.wikipedia.org/wiki/HSL_and_HSV).

I also added a "within" matcher for both HSLColor and HSVColor that will check if the (floating point) color components are within a certain error.

And tests.

### CODE REVIEW

File: packages/flutter/lib/src/painting/colors.dart [hunk 1]
```diff
@@ -49,9 +62,10 @@ class HSVColor {
     } else if (max == blue) {
       hue = 60.0 * (((red - green) / delta) + 4);
     }
-    
+
     /// fix hue to 0.0 when red == green == blue.
     hue = hue.isNaN ? 0.0 : hue;
+    hue = hue % 360.0;
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/colors.dart]: how can it be out of range?

File: packages/flutter/lib/src/painting/colors.dart [hunk 2]
```diff
@@ -169,10 +185,10 @@ class HSVColor {
     if (b == null)
       return a._scaleAlpha(1.0 - t);
     return new HSVColor.fromAHSV(
-      lerpDouble(a.alpha, b.alpha, t),
-      lerpDouble(a.hue, b.hue, t),
-      lerpDouble(a.saturation, b.saturation, t),
-      lerpDouble(a.value, b.value, t),
+      lerpDouble(a.alpha, b.alpha, t).clamp(0.0, 1.0),
+      lerpDouble(a.hue, b.hue, t).clamp(0.0, 360.0),
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/colors.dart]: why not mod-360?

File: packages/flutter/lib/src/painting/colors.dart [hunk 3]
```diff
@@ -196,6 +212,211 @@ class HSVColor {
...
+      hue = 60.0 * (((blue - red) / delta) + 2);
+    } else if (max == blue) {
+      hue = 60.0 * (((red - green) / delta) + 4);
+    }
+
+    /// fix hue to 0.0 when red == green == blue.
+    hue = hue.isNaN ? 0.0 : hue;
+    final double saturation = max == 0.0 ? 0.0 : delta / max;
+
+    return new HSLColor.fromAHSL(alpha, hue, saturation, lightness);
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/colors.dart]: this is a lot of duplicated code... can we share it somehow?

File: packages/flutter/lib/src/painting/colors.dart [hunk 4]
```diff
@@ -196,6 +212,211 @@ class HSVColor {
...
+  final double alpha;
+
+  /// Hue, from 0.0 to 360.0.
+  final double hue;
+
+  /// Saturation, from 0.0 to 1.0.
+  final double saturation;
+
+  /// Value, from 0.0 to 1.0.
+  final double lightness;
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/colors.dart]: the docs are definitely wrong here.

We should define what "lightness" and "value" are in the respective dartdocs.

File: packages/flutter/lib/src/painting/colors.dart [hunk 5]
```diff
@@ -196,6 +212,211 @@ class HSVColor {
...
+    return new HSLColor.fromAHSL(alpha, hue, saturation, lightness);
+  }
+
+  /// Returns a copy of this color with the saturation parameter replaced with the given lightness.
+  HSLColor withSaturation(double saturation) {
+    return new HSLColor.fromAHSL(alpha, hue, saturation, lightness);
+  }
+
+  /// Returns a copy of this color with the lightness parameter replaced with the given lightness.
+  HSLColor withValue(double lightness) {
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/colors.dart]: i think you may have been overly aggressive in your s/value/lightness/...

Comment (Hixie) [reply packages/flutter/lib/src/painting/colors.dart]: (e.g. check the dartdocs for all these methods)

**Comment by Hixie** [comment]
What's the motivation for having both HSV and HSL?

Can we put the lightness as a getter on HSVColor? Or is the functionality substantially different? (e.g. does it lerp differently?)

**Comment by gspencergoog** [comment]
One motivation is that the algorithm for selecting palette colors in the [palette generator](https://github.com/flutter/flutter/pull/18436) is predicated on the HSL colorspace. Another is that having an alternate to HSV that is a perceptual colorspace is desirable because it operates differently with respect to the perceptual mapping of brightness: HSV is modeled on how pigments combine, and HSL tries to follow more closely how humans perceive it, and is modeled more on how colored light combines perceptually.

They are just different colorspaces (see the [Wikipedia link](https://en.wikipedia.org/wiki/HSL_and_HSV)).  I tried to make a superclass for "cylindrical" colorspace colors, but  it would have just basically been an interface because I would have to override each of the implementation functions for `lerp`, `toColor`, etc., and the accessors for the luminosity (value/lightness) had different names.

I try again to combine them, although I think I will make it a private thing, since they really are apples and (apple-shaped) oranges, and there aren't too many other cylindrical colorspaces. Maybe aggregation as a private helper would work better.

File: packages/flutter/lib/src/painting/colors.dart [hunk 6]
```diff
@@ -196,6 +212,211 @@ class HSVColor {
...
+      hue = 60.0 * (((green - blue) / delta) % 6);
+    } else if (max == green) {
+      hue = 60.0 * (((blue - red) / delta) + 2);
+    } else if (max == blue) {
+      hue = 60.0 * (((red - green) / delta) + 4);
+    }
+
+    /// fix hue to 0.0 when red == green == blue.
+    hue = hue.isNaN ? 0.0 : hue;
+    final double saturation = max == 0.0 ? 0.0 : delta / max;
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/colors.dart]: according to the wikipedia article, Saturation is different for HSL and HSV, but you compute it the same in both cases.

**Comment by Hixie** [comment]
Mostly it's the code duplication that's sad, though I agree that it's not clear how to avoid it without sacrificing memory or performance.

**Comment by gspencergoog** [comment]
OK, I was able to at least factor out and share the hue parts of the code (which had the big conditional stacks), but doing anything more extensive is hard, mainly because these are immutable, so anything that wants to change something has to make a new one, which means either passing in a factory function to the "common" code, or doing a lot of type checking, neither of which seemed like fun (or a good idea for this use case).

I did fix the saturation calculation. I also added a bunch of better tests.

File: packages/flutter/lib/src/painting/colors.dart [hunk 7]
```diff
@@ -193,7 +227,177 @@ class HSVColor {
...
+
+  /// Hue, from 0.0 to 360.0. Describes which color of the spectrum is
+  /// represented. A value of 0.0 represents red, as does 360.0. Values in
+  /// between go through all the hues representable in RGB.
+  final double hue;
+
+  /// Saturation, from 0.0 to 1.0. This describes how colorful the color is.
+  /// 0.0 implies a shade of grey, and 1.0 implies a color as vibrant as that
+  /// hue gets.
+  final double saturation;
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/colors.dart]: this is exactly the same definition as for HSV, but the math is different, which doesn't make sense to me.

**Comment by Hixie** [comment]
LGTM with two documentary additions: the `saturation` docs should explain how they are different from each other and caution the reader as to the fact that they are different, and the HSLColor and HSVColor class docs should point to each other and provide motivation for when you'd use one or the other.

**Comment by gspencergoog** [comment]
OK, I think I've added the appropriate documentation.  Let me know if you want more/different docs, and I'll add it.

**Pull request merged**